docs/source/advanced_usage.rst

```diff
@@ -425,6 +425,141 @@ We can run this program a few times to see what we get in the readout register `
+        program = Program()
+        program.if_then(measures[0], Program(X(qubits[0]), X(qubits[1])))
+        return program
+
+Enforcing a sentinel condition
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Next, we'll construct the part of our program that enforces the sentinel condition. In this case, we'll end the program
+if the given memory reference is 0, otherwise we'll want to reset the state of our qubits and jump back to the
+beginning of the program. We'll construct the branch that ends the program using Quil's ``Halt`` instruction, and we'll
+accept the alternative branch as an argument to our function and pass it in the next step:
+
+.. testcode:: sentinel-based-loop
+
+    def enforce_sentinel(mem_ref: MemoryReference, else_program: Program) -> Program:
+        """Ends the program if mem_ref is 0, otherwise executes else_program."""
+        program = Program()
+        program.if_then(mem_ref, else_program, Program(Halt()))
+        return program
+
+Putting it all together
+^^^^^^^^^^^^^^^^^^^^^^^
+
+With each component of our program ready, we just need to compose all the pieces:
+
+.. testcode:: sentinel-based-loop
+
+    def sentinel_program(qubits: Tuple[Qubit, Qubit]) -> Program:
+        # Create a label to reference the start of the program
+        start_label = Label("start-loop")
+        
+        # Use the label to create a jump target at the beginning of the program
+        program = Program(JumpTarget(start_label))
+        
+        # Declare a register to measure the qubits into
+        measures = program.declare("measures", "BIT", 2)
+
+        # Add the loop body to our program
+        program += body(qubits, measures)
+        
+        # If the sentinel condition isn't met, then we: 
+        reset = Program(
+            # Reset the state of our qubits
+            reset_bell_state(qubits, measures),
+            # Jump back to the start of the program
+            Jump(start_label)
+        )
+        
+        # Finally, if both Qubits measured to 0 (our sentinel), then we want to end the program
+        # Otherwise, we try again.
+        program += enforce_sentinel(measures[0], reset)
+        
+        # We used pyQuil to construct some of the branches for us, those methods use label placeholders
+        # to avoid conflicts with existing labels in the program, so we resolve those placeholders here.
+        program.resolve_label_placeholders()
+        return program
+
+Testing our program
+^^^^^^^^^^^^^^^^^^^
+
+Now that we have our program, let's test it out. We'll use the ``sentinel_program`` function to construct the program
+and run it against a QVM for 1000 shots. We'll use numpy to assert that the measures register contains only 0s. Over
+1000 trials, this result would be improbable if our program didn't work as intended.
+
+.. testcode:: sentinel-based-loop
+
+    qubits = (Qubit(0), Qubit(1))
+
+    qc = get_qc("2q-qvm")
+    program = sentinel_program(qubits)
```

Comment:
Might be helpful (or confusing? idk) to note the output of `print(program)` (blank lines added)

```
DECLARE measures BIT[2]

LABEL @start-loop
H 0
CNOT 0 1
MEASURE 0 measures[0]
MEASURE 1 measures[1]
JUMP-WHEN @THEN_0 measures[0]
HALT
JUMP @END_0

LABEL @THEN_0
JUMP-WHEN @THEN_1 measures[0]
JUMP @END_1

LABEL @THEN_1
X 0
X 1

LABEL @END_1
JUMP @start-loop

LABEL @END_0
```

Reply:
Yeah, I thought it might be a bit noisy, but I'll go ahead and print it in the docs.